language-plutus-core/plutus-ir/Language/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -0,0 +1,615 @@
+
+-- [infer| G !- arg :: k]    [check| G !- pat :: (k -> *) -> k -> *]    pat ~>? vPat    arg ~>? vArg
+-- [check| G !- term : NORM (vPat (\(a :: k) -> ifix vPat a) vArg)]
+-- -------------------------------------------------------------------------------------------------
+-- [infer| G !- iwrap pat arg term : ifix vPat vArg]
+inferTypeM (IWrap ann pat arg term) = do
+    k <- inferKindM arg
+    checkKindOfPatternFunctorM ann pat k
+    vPat <- normalizeTypeM $ void pat
+    vArg <- normalizeTypeM $ void arg
+    checkTypeM ann term =<< unfoldFixOf vPat vArg k
+    pure $ TyIFix () <$> vPat <*> vArg
+
+-- [infer| G !- term : ifix vPat vArg]    [infer| G !- vArg :: k]
+-- -----------------------------------------------------------------------
+-- [infer| G !- unwrap term : NORM (vPat (\(a :: k) -> ifix vPat a) vArg)]
+inferTypeM (Unwrap ann term)        = do
+    vTermTy <- inferTypeM term
+    case unNormalized vTermTy of
+        TyIFix _ vPat vArg -> do
+            k <- inferKindM $ ann <$ vArg
+            -- Subparts of a normalized type, so normalized.
+            unfoldFixOf (Normalized vPat) (Normalized vArg) k
+        _                  -> throwError (TypeMismatch ann (void term) (TyIFix () dummyType dummyType) vTermTy)
+-- [check| G !- ty :: *]    ty ~>? vTy
+-- -----------------------------------
+-- [infer| G !- error ty : vTy]
+inferTypeM (Error ann ty)           = do
+    checkKindM ann ty $ Type ()
+    normalizeTypeM $ void ty
+
+inferTypeM (Let ann1 recurs bs inTerm) = do
+    tyInTerm <- case recurs of
+        -- [check| G !- ty :: *]  [infer| G,withBindsLinear(b) !- inT : ty] (forall b in bs. [checkWellFormed| Glinear !- b]) 
+        -- -------------------------------------------------
+        -- [infer| G !- (let (nonrec) bs in inT) :: ty
+        NonRec ->
+            foldr
+               (\b acc -> do
+                   checkWellformBind NonRec b
+                   -- linear scope: the new identifiers will be visible in the next binding
+                   withBind b acc)
+               (inferTypeM inTerm)
+               bs
+        -- [check| G !- ty :: *]  G'=G,withBinds(bs)  [infer| G' !- inT : ty] (forall b in bs. [checkWellFormed| G' !- b])
+        -- -------------------------------------------------
+        -- [infer| G !- (let (rec) bs in inT) :: ty
+        Rec -> do
+            -- Check that there are no conflicting new identifiers (same-name)
+            -- that are introduced by the bindings of this the letrec
+            -- separately check the VarTypes environment for conflicts
+            -- TODO: move these 2 checks to a different pass, e.g. in renaming phase
+            assertNoRecConflict $ foldMap bindingNewIds bs
+            -- separately check the VarkInds environment for conflicts
+            assertNoRecConflict . mapMaybe bindingNewTyId $ toList bs
+            -- now with all bindings in the env, check for wellformedness
+            withBinds bs $ do
```

Comment:
Worth a note: this would allow later bindings to shadow earlier ones, but that doesn't matter because we should have globally unique uniques at this point.